diff --git a/src/com/android/providers/media/MediaProvider.java b/src/com/android/providers/media/MediaProvider.java
index 9d4c036..2ff5404 100755
--- a/src/com/android/providers/media/MediaProvider.java
+++ b/src/com/android/providers/media/MediaProvider.java
@@ -5031,7 +5031,7 @@ public class MediaProvider extends ContentProvider {
             // request a reasonably sized output image
             final Resources r = getContext().getResources();
             final int maximumThumbSize = r.getDimensionPixelSize(R.dimen.maximum_thumb_size);
-            while (opts.outHeight > maximumThumbSize || opts.outWidth > maximumThumbSize) {
+            while (opts.outHeight > 1024 || opts.outWidth > 1024) {
                 opts.outHeight /= 2;
                 opts.outWidth /= 2;
                 opts.inSampleSize *= 2;
